[compiler] Remove ConstAPI struct

staticfloat opened the pull request:
`ConstAPI` predates the effect analysis framework, and was relying on an
ad-hoc implementation of effect analysis that only worked for functions
of less than 15 statements.  As a result, it wasn't used much anymore,
but it could cause trouble with external `AbstractInterpreters` that may
want to cache additional code metadata within `CodeInstance`s.

This commit simply removes this, expecting that the existing effect
analysis will pick up the slack.

staticfloat committed:
[compiler] Remove `ConstAPI` struct

`ConstAPI` predates the effect analysis framework, and was relying on an
ad-hoc implementation of effect analysis that only worked for functions
of less than 15 statements.  As a result, it wasn't used much anymore,
but it could cause trouble with external `AbstractInterpreters` that may
want to cache additional code metadata within `CodeInstance`s.

This commit simply removes this, expecting that the existing effect
analysis will pick up the slack.

staticfloat committed:
Add short-circuit `has_offset_axes(::Array)`

This ensures that `Base.require_one_based_indexing()` is always fully
eliminated, since previously `has_offset_axes(::Array)` is
effect-tainted as non-consistent due to calculating `size(::Array)`.
This dodges the effect taint by adding this short-circuit, fixing broken
tests due to using effects analysis more extensively.

aviatesk commented on base/compiler/typeinfer.jl:
```diff
@@ -300,9 +292,10 @@ function CodeInstance(
     @assert !(result_type isa LimitedAccuracy)
-    if inferred_result isa ConstAPI
+
+    if isa(result_type, Const) && is_total(result.ipo_effects)
```
I think we only require `is_foldable && is_nothrow` here?

aviatesk replied:
And I think we should also check `is_inlineable_constant(result_type.value)` here.

staticfloat replied:
> I think we only require is_foldable && is_nothrow here?

I think `is_total()` is defined as ` is_foldable` && `is_nothrow`?

aviatesk replied:
> I think `is_total()` is defined as `is_foldable` && `is_nothrow`?

Yes, I just remembered it. Perhaps we should have revised its definition as we added the additional effects. "is total" is now a much stronger condition that covers the other guarantees, such as `is_notaskstate` and `is_inaccessiblememonly`. So be safe to use `is_foldable && is_nothrow` here and in the inlining handling.

staticfloat replied:
So you're saying that `is_total()` may change its definition in the future?

aviatesk commented on base/compiler/typeinfer.jl:
```diff
@@ -379,7 +372,7 @@ function transform_result_for_cache(interp::AbstractInterpreter,
     end
     # The global cache can only handle objects that codegen understands
-    if !isa(inferred_result, Union{CodeInfo, Vector{UInt8}, ConstAPI})
+    if !isa(inferred_result, Union{CodeInfo, Vector{UInt8}})
```
```suggestion
    if !isa(inferred_result, MaybeCompressed)
```

N5N3 committed:
Avoid split `y::Union` when comparing with a forall var.

If `y` has no free typevar. Then `x <:  y` wont cause any env change. Thus there's no need to split `y`.

staticfloat committed:
Address review comments

staticfloat committed:
Move coverage effects analysis into inference

Code coverage is implemented as an optimization pass that inserts an SSA
IR node during optimization, however effects analysis is perfomed during
inference, and so the desired effects tainting that should occur from
insertion of the code coverage IR node is ignored.

This teaches `InferenceState` to taint itself with the effects flag
 `effects_free = ALWAYS_FALSE` when coverage is applicable to this
inference state's method.

staticfloat commented:
I've added a change to try and fix the `.cov` test failure, and while it improves the IR, we're still not getting a `.cov` file.

staticfloat committed:
Try to fix the tests

We just assume that you don't want code coverage if you're using this
constructor. This may not always be true.

staticfloat commented:
I believe I have fixed the last bit of tests related to code coverage.

Code coverage is implemented by inserting a special code coverage IR node into the IR, however it is inserted at optimization time, which is after inference is finished assigning effects to return types.  To fix this, we moved the code coverage check (including scoped coverage selection) into inference time, mark each `InferenceFrame` as needing a coverage node insertion or not, and then taint the effects as "not effect free".  This correctly prevents the coverage nodes from being dropped by later optimization passes, and brings back the coverage we all know and love.

staticfloat commented:
One remaining question is the second constructor to `OptimizationState`; it does not contain a reference to its `InferenceState`, so we can't receive the `insert_coverage` boolean; because the core compiler does not use this constructor, I assume it's okay to just hardcode `insert_coverage` as `false` here; but if not, I'd appreciate guidance on what to do here instead.

oscardssmith committed:
Merge branch 'master' into sf/ConstAPI_must_die
